File: leet-code/783. Minimum Distance Between BST Nodes/MinimumDistanceBetweenBSTNodes783.java
package com.example;

public class MinimumDistanceBetweenBSTNodes783 {
    private int minDiff = Integer.MAX_VALUE;

    // TreeNode 최대값이 100000 이라 오버플로우 방지
    private int prev = Integer.MIN_VALUE + 100000;

    public int minDiffInBST(TreeNode root) {
        if (root == null) return 0;
        minDiffInBST(root.left);
        minDiff = Math.min(minDiff, Math.abs(prev - root.val));
        prev = root.val;
        minDiffInBST(root.right);
        return minDiff;
    }
}

File: leet-code/783. Minimum Distance Between BST Nodes/MinimumDistanceBetweenBSTNodes783.kt
package com.example

import kotlin.math.min

class MinimumDistanceBetweenBSTNodes783 {
    private var minDistance = 100001
    private var previous = -100001

    fun minDiffInBST(root: TreeNode?): Int {
        inorderTraverse(root)
        return minDistance
    }

    private fun inorderTraverse(root: TreeNode?) {
        if (root!!.left != null) {
            minDiffInBST(root.left)
        }
        minDistance = min(minDistance, root.`val` - previous)
        previous = root.`val`
        if (root.right != null) {
            minDiffInBST(root.right)
        }
    }
}
